numpy/core/tests/test_numeric.py

```diff
@@ -991,9 +998,103 @@ class C(np.ndarray):
                         assert_(type(nzx_i) is np.ndarray)
                         assert_(nzx_i.flags.writeable)
 
-    # Tests that the array method
-    # call works
+    def test_count_nonzero_axis(self):
+        # Basic check of functionality
+        m = np.array([[0, 1, 7, 0, 0], [3, 0, 0, 2, 19]])
+
+        expected = np.array([1, 1, 1, 1, 1])
+        assert_equal(np.count_nonzero(m, axis=0), expected)
+
+        expected = np.array([2, 3])
+        assert_equal(np.count_nonzero(m, axis=1), expected)
+
+        self.assertRaises(ValueError, np.count_nonzero, m, axis=(1, 1))
+        self.assertRaises(TypeError, np.count_nonzero, m, axis='foo')
+        self.assertRaises(ValueError, np.count_nonzero, m, axis=3)
+        self.assertRaises(TypeError, np.count_nonzero,
+                          m, axis=np.array([[1], [2]]))
+
+    def test_count_nonzero_axis_all_dtypes(self):
+        # More thorough test that the axis argument is respected
+        # for all dtypes and responds correctly when presented with
+        # either integer or tuple arguments for axis
+        msg = "Mismatch for dtype: %s"
+
+        for dt in np.typecodes['All']:
+            err_msg = msg % (np.dtype(dt).name,)
+
+            if dt != 'V':
+                if dt != 'M':
+                    m = np.zeros((3, 3), dtype=dt)
+                    n = np.ones(1, dtype=dt)
+
+                    m[0, 0] = n[0]
+                    m[1, 0] = n[0]
+
+                else:  # np.zeros doesn't work for np.datetime64
+                    from datetime import datetime
+
+                    m = np.array([datetime.fromtimestamp(0)] * 9)
```

Comment:
I don't think any datetimes should evaluate as zero. So you may have encountered a bug in the existing behavior. See [here](https://mail.python.org/pipermail/python-ideas/2014-March/026647.html) for more discussion on this related points.

One other problematic aspect -- `fromtimestamp` returns a *local time* from a POSIX timestamp. So I think this test might only pass if run on a computer whose timezone is set to UTC. Even if not, better to use `np.datetime64` directly to be safe.